README.md
# Ini

[![GoDoc](https://godoc.org/github.com/Thomasdezeeuw/ini?status.svg)](https://godoc.org/github.com/Thomasdezeeuw/ini)
[![Build Status](https://img.shields.io/travis/Thomasdezeeuw/ini.svg)](https://travis-ci.org/Thomasdezeeuw/ini)
[![Coverage Status](https://coveralls.io/repos/Thomasdezeeuw/ini/badge.svg?branch=master&service=github)](https://coveralls.io/github/Thomasdezeeuw/ini?branch=master)
[![Go Report Card](https://goreportcard.com/badge/github.com/Thomasdezeeuw/ini)](https://goreportcard.com/report/github.com/Thomasdezeeuw/ini)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/Thomasdezeeuw/ini/blob/master/LICENSE)

Ini is a [Go](https://golang.org/) package for for parsing ini (or cfg) files.
See [Godoc](https://godoc.org/github.com/Thomasdezeeuw/ini) for the API.

## Stable

The api is now stable and will remain stable until version 2 is released. To
prove how stable `ini.Parse` is, it battled [go-fuzz](https://github.com/dvyukov/go-fuzz)
for 10 hours, processing over 1.1 billion randomized inputs and won:

```
2015/07/21 23:58:54 slaves: 8, corpus: 611 (3h8m ago), crashers: 0, restarts: 1/
10000, execs: 1100121441 (30462/sec), cover: 493, uptime: 10h1m
```

## Installation

Run the following line to install.

```bash
$ go get github.com/Thomasdezeeuw/ini
```

## Examples

See the [_examples](https://github.com/Thomasdezeeuw/ini/tree/master/_examples)
directory for a multiple examples.

## License

Licensed under the MIT license, copyright (C) Thomas de Zeeuw. See `LICENSE` for
more information.


CHANGELOG.md
# Changelog

## v0.2

- Fixed various type-o's in documentation.
- Renamed `IsSynthaxError` to `IsSyntaxError`.
- Fixed type-o in syntax error message.

## v0.1

 - **Ignore goveralls errors in travis build** (#c57cb74) by *Thomas de Zeeuw*, on *02 Apr 2016 12:31:57 UTC*. <br/>
*The service is to flaky to fail the build for*.
 - **Simplify test code** (#5257b94) by *Thomas de Zeeuw*, on *02 Apr 2016 11:19:24 UTC*. <br/>
*Found using https://github.com/dominikh/go-simple*.
 - **Remove unnecessary type conversions** (#88d4b5f) by *Thomas de Zeeuw*, on *15 Mar 2016 19:12:29 UTC*. <br/>
*Found using https://github.com/mdempsky/unconvert*.
 - **Fix two type-o's** (#57a5d6e) by *Thomas de Zeeuw*, on *09 Feb 2016 18:27:14 UTC*. <br/>
*Found with Misspell*.
 - **Add MIT license badge** (#c17a421) by *Thomas de Zeeuw*, on *09 Feb 2016 18:26:22 UTC*.
 - **Add Go Report Card** (#442052f) by *Thomas de Zeeuw*, on *09 Feb 2016 18:26:13 UTC*.
 - **Use shields.io badge for travis** (#d0acfbc) by *Thomas de Zeeuw*, on *09 Feb 2016 18:26:02 UTC*. <br/>
*Not as pixelated, https://shields.io*.
 - **Remove todo** (#045541f) by *Thomas de Zeeuw*, on *26 Jan 2016 15:25:10 UTC*. <br/>
*These have been converted into issues on GitHub*.
 - **Happy 2016** (#53efc69) by *Thomas de Zeeuw*, on *26 Jan 2016 11:51:38 UTC*.
 - **Make link to examples directory in ready** (#1f522b4) by *Thomas de Zeeuw*, on *10 Dec 2015 11:59:23 UTC*. <br/>
*Closes #17*.
 - **Display full utf-8 character in error messages** (#df315a8) by *Thomas de Zeeuw*, on *08 Dec 2015 20:27:17 UTC*. <br/>
*Closes #14*.
 - **Support simple custom types** (#b15cb7e) by *Thomas de Zeeuw*, on *07 Dec 2015 19:18:21 UTC*. <br/>
*Instead of getting the type of a reflect value get the kind (the go type) and use that in determining how to set the value.  Updates #16*.
 - **Allow # as a start of a comment** (#a98383c) by *Thomas de Zeeuw*, on *07 Dec 2015 19:00:25 UTC*. <br/>
*Closes #13*.
 - **Fix a lot of type-o's** (#6c1de5b) by *Thomas de Zeeuw*, on *07 Dec 2015 15:51:21 UTC*. <br/>
*seperator -> separator. qoute -> quote.  Closes #12*.
 - **Don't trim whitespace for quoted values** (#872a7a4) by *Thomas de Zeeuw*, on *07 Dec 2015 15:40:55 UTC*. <br/>
*Closed #11*.
 - **Update getNameParts doc** (#474ac87) by *Thomas de Zeeuw*, on *07 Dec 2015 15:39:23 UTC*. <br/>
*Type-o and start the line with a capital*.
 - **Small code cleanup** (#1327c25) by *Thomas de Zeeuw*, on *06 Dec 2015 15:52:37 UTC*. <br/>
*Move getConfigSectionsAlpha after first use*.
 - **Doc getSectionKeysAlpha** (#bcb8ca1) by *Thomas de Zeeuw*, on *06 Dec 2015 15:52:25 UTC*.
 - **Update examples in ready** (#759bdd9) by *Thomas de Zeeuw*, on *06 Dec 2015 15:37:46 UTC*.
 - **Create new Section type** (#8c9c321) by *Thomas de Zeeuw*, on *06 Dec 2015 15:34:23 UTC*. <br/>
*Closes #6*.
 - **Doc tags in Config.Decode** (#212ad3d) by *Thomas de Zeeuw*, on *06 Dec 2015 15:29:47 UTC*. <br/>
*Updates #4*.
 - **Use strconv.ParseBool for converting bools** (#8049e5e) by *Thomas de Zeeuw*, on *06 Dec 2015 15:22:53 UTC*.
 - **Add examples for using tags** (#3e79ab0) by *Thomas de Zeeuw*, on *06 Dec 2015 15:13:25 UTC*. <br/>
*Closes #7.  Updates #4*.
 - **Add decode example** (#1d6e07b) by *Thomas de Zeeuw*, on *06 Dec 2015 15:06:11 UTC*. <br/>
*Updates #7*.
 - **Fix simple log** (#a39522b) by *Thomas de Zeeuw*, on *06 Dec 2015 15:05:55 UTC*.
 - **Fix type-o in license info in source files** (#f6a6d90) by *Thomas de Zeeuw*, on *06 Dec 2015 14:57:47 UTC*.
 - **Create new examples** (#a7dcb87) by *Thomas de Zeeuw*, on *06 Dec 2015 14:54:58 UTC*. <br/>
*Starting with a simple example in the new `_examples` directory.  Updates #7*.
 - **Always check errors** (#a3d4201) by *Thomas de Zeeuw*, on *04 Dec 2015 18:42:39 UTC*. <br/>
*Even in examples*.
 - **Small doc update** (#4746a38) by *Thomas de Zeeuw*, on *04 Dec 2015 18:03:50 UTC*.
 - **Simplify parseSection code** (#ba51dac) by *Thomas de Zeeuw*, on *04 Dec 2015 18:03:35 UTC*.
 - **Remove TestPossibleQoute** (#3c72634) by *Thomas de Zeeuw*, on *21 Nov 2015 19:45:48 UTC*.
 - **Always quote keys and values in Config.String/Bytes** (#9f43ad8) by *Thomas de Zeeuw*, on *21 Nov 2015 19:44:42 UTC*.
 - **Split TestParse tests** (#07a5b27) by *Thomas de Zeeuw*, on *21 Nov 2015 19:38:04 UTC*.
 - **Fix test error output** (#b0e3cce) by *Thomas de Zeeuw*, on *21 Nov 2015 17:50:06 UTC*.
 - **Small code cleanup** (#f5f96ab) by *Thomas de Zeeuw*, on *21 Nov 2015 17:43:57 UTC*.
 - **Add coverage badge to readme** (#8d832c9) by *Thomas de Zeeuw*, on *21 Nov 2015 17:01:36 UTC*.
 - **Clean travis install commands** (#d25f35f) by *Thomas de Zeeuw*, on *21 Nov 2015 17:01:10 UTC*. <br/>
*Drop `get ./…`, no dependencies. Comment out `github.com/fzipp/gocyclo`, no used, but it should be. Also don’t update any of them*.
 - **Fix data race with testing with coverage** (#3b7bb4a) by *Thomas de Zeeuw*, on *21 Nov 2015 16:55:52 UTC*. <br/>
*Covermode count creates data races, atomic fixes this*.
 - **Add coveralls.io** (#b6974ed) by *Thomas de Zeeuw*, on *21 Nov 2015 16:24:21 UTC*.
 - **Remove glint from Travis** (#687a142) by *Thomas de Zeeuw*, on *04 Nov 2015 14:59:39 UTC*. <br/>
*Golint never exits with status code 1, so we’ll never see if it has any comments*.
 - **Fix travis.yml file** (#b32db3d) by *Thomas de Zeeuw*, on *04 Nov 2015 13:43:30 UTC*. <br/>
*Have to use spaces, tab aren’t allowed*.
 - **Travis more advanced testing** (#4f3083f) by *Thomas de Zeeuw*, on *04 Nov 2015 13:40:49 UTC*. <br/>
*Run gofmt simplify, go vet, deadcode, golint and go test with the race detector enabled.  Todo enable gocyclo*.
 - **Travis add OS X as a os to run tests on** (#898c968) by *Thomas de Zeeuw*, on *04 Nov 2015 13:39:18 UTC*.
 - **Remove unused code** (#e45e98e) by *Thomas de Zeeuw*, on *04 Nov 2015 13:34:34 UTC*.
 - **Fix license type-o** (#a1050b0) by *Thomas de Zeeuw*, on *15 Oct 2015 20:36:04 UTC*.

scan.go
// Copyright (C) 2015 Thomas de Zeeuw.
//
// Licensed onder the MIT license that can be found in the LICENSE file.

package ini

import (
	"errors"
	"io"
	"reflect"
	"strconv"
	"strings"
	"time"
)

var timeFormats = []string{"2006-01-02", "2006-01-02 15:04",
	"2006-01-02 15:04:05"}

var (
	typeString   = reflect.TypeOf("")
	typeBool     = reflect.TypeOf(true)
	typeInt      = reflect.TypeOf(int(1))
	typeInt8     = reflect.TypeOf(int8(1))
	typeInt16    = reflect.TypeOf(int16(1))
	typeInt32    = reflect.TypeOf(int32(1))
	typeInt64    = reflect.TypeOf(int64(1))
	typeUint     = reflect.TypeOf(uint(1))
	typeUint8    = reflect.TypeOf(uint8(1))
	typeUint16   = reflect.TypeOf(uint16(1))
	typeUint32   = reflect.TypeOf(uint32(1))
	typeUint64   = reflect.TypeOf(uint64(1))
	typeFloat32  = reflect.TypeOf(float32(1.0))
	typeFloat64  = reflect.TypeOf(float64(1.0))
	typeDuration = reflect.TypeOf(time.Nanosecond)
	typeTime     = reflect.TypeOf(time.Time{})
)

// Scan scans a configuration into a struct or map, see `Config.Scan`.
func Scan(r io.Reader, dst interface{}) error {
	c, err := Parse(r)
	if err != nil {
		return err
	}
	return c.Scan(dst)
}

// ScanInto scans a single configuration value into a variable.
func ScanInto(value string, dst interface{}) error {
	valuePtr := reflect.ValueOf(dst)
	v := reflect.Indirect(valuePtr)

	// If it's not a pointer we can't change the original value.
	if valuePtr.Kind() != reflect.Ptr {
		return errors.New("ini: ScanInto requires a pointer to a destination value")
	} else if !v.IsValid() || !v.CanSet() {
		return errors.New("ini: can't change value of destination value")
	}

	return setReflectValue(&v, value)
}

func setReflectValue(keyValue *reflect.Value, value string) error {
	if keyValue.Kind() == reflect.Slice {
		return setSlice(keyValue, value)
	}

	switch keyValue.Type() {
	case typeString:
		keyValue.SetString(value)
	case typeBool:
		setBool(keyValue, value)
	case typeInt, typeInt8, typeInt16, typeInt32, typeInt64:
		return setInt(keyValue, value)
	case typeUint, typeUint8, typeUint16, typeUint32, typeUint64:
		return setUint(keyValue, value)
	case typeFloat32, typeFloat64:
		return setFloat(keyValue, value)
	case typeDuration:
		return setDuration(keyValue, value)
	case typeTime:
		return setTime(keyValue, value)
	}

	return nil
}

// SetSlice sets a slice of reflected value.
func setSlice(keyValue *reflect.Value, value string) error {
	values := getValues(value)

	// Type of the slice elements.
	switch keyValue.Type().Elem() {
	case typeString:
		keyValue.Set(reflect.ValueOf(values))
	case typeBool:
		return setBools(keyValue, values)
	case typeInt:
		return setInts(keyValue, values)
	case typeInt8:
		return setInt8s(keyValue, values)
	case typeInt16:
		return setInt16s(keyValue, values)
	case typeInt32:
		return setInt32s(keyValue, values)
	case typeInt64:
		return setInt64s(keyValue, values)
	case typeUint:
		return setUints(keyValue, values)
	case typeUint8:
		return setUint8s(keyValue, values)
	case typeUint16:
		return setUint16s(keyValue, values)
	case typeUint32:
		return setUint32s(keyValue, values)
	case typeUint64:
		return setUint64s(keyValue, values)
	case typeFloat32:
		return setFloat32s(keyValue, values)
	case typeFloat64:
		return setFloat64s(keyValue, values)
	case typeDuration:
		return setDurations(keyValue, values)
	case typeTime:
		return setTimes(keyValue, values)
	}

	return nil
}

func setBools(keyValue *reflect.Value, values []string) error {
	var bs = make([]bool, len(values))
	for i, value := range values {
		bValue := reflect.Indirect(reflect.ValueOf(&bs[i]))
		setBool(&bValue, value)
	}
	bsValue := reflect.ValueOf(bs)
	keyValue.Set(bsValue)
	return nil
}

// Returns true on "1" and "true" (case insensitive), anything returns false.
func setBool(keyValue *reflect.Value, value string) {
	var b bool
	value = strings.TrimSpace(value)
	if value == "1" || strings.ToLower(value) == "true" {
		b = true
	}
	keyValue.SetBool(b)
}

func setInts(keyValue *reflect.Value, values []string) error {
	var ns = make([]int, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setInt(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setInt8s(keyValue *reflect.Value, values []string) error {
	var ns = make([]int8, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setInt(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setInt16s(keyValue *reflect.Value, values []string) error {
	var ns = make([]int16, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setInt(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setInt32s(keyValue *reflect.Value, values []string) error {
	var ns = make([]int32, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setInt(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setInt64s(keyValue *reflect.Value, values []string) error {
	var ns = make([]int64, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setInt(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setInt(keyValue *reflect.Value, value string) error {
	n, err := strconv.ParseInt(value, 10, 64)
	if err != nil {
		return createCovertionError(value, keyValue.Kind().String())
	}
	n64 := int64(n)

	if keyValue.OverflowInt(n64) {
		return createOverflowError(value, keyValue.Kind().String())
	}

	keyValue.SetInt(n64)
	return nil
}

func setUints(keyValue *reflect.Value, values []string) error {
	var ns = make([]uint, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setUint(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setUint8s(keyValue *reflect.Value, values []string) error {
	var ns = make([]uint8, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setUint(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setUint16s(keyValue *reflect.Value, values []string) error {
	var ns = make([]uint16, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setUint(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setUint32s(keyValue *reflect.Value, values []string) error {
	var ns = make([]uint32, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setUint(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setUint64s(keyValue *reflect.Value, values []string) error {
	var ns = make([]uint64, len(values))
	for i, value := range values {
		nValue := reflect.Indirect(reflect.ValueOf(&ns[i]))
		if err := setUint(&nValue, value); err != nil {
			return err
		}
	}
	nsValue := reflect.ValueOf(ns)
	keyValue.Set(nsValue)
	return nil
}

func setUint(keyValue *reflect.Value, value string) error {
	n, err := strconv.ParseUint(value, 10, 64)
	if err != nil {
		return createCovertionError(value, keyValue.Kind().String())
	}
	nu64 := uint64(n)

	if keyValue.OverflowUint(nu64) {
		return createOverflowError(value, keyValue.Kind().String())
	}

	keyValue.SetUint(nu64)
	return nil
}

func setFloat32s(keyValue *reflect.Value, values []string) error {
	var fs = make([]float32, len(values))
	for i, value := range values {
		fValue := reflect.Indirect(reflect.ValueOf(&fs[i]))
		if err := setFloat(&fValue, value); err != nil {
			return err
		}
	}
	fsValue := reflect.ValueOf(fs)
	keyValue.Set(fsValue)
	return nil
}

func setFloat64s(keyValue *reflect.Value, values []string) error {
	var fs = make([]float64, len(values))
	for i, value := range values {
		fValue := reflect.Indirect(reflect.ValueOf(&fs[i]))
		if err := setFloat(&fValue, value); err != nil {
			return err
		}
	}
	fsValue := reflect.ValueOf(fs)
	keyValue.Set(fsValue)
	return nil
}

func setFloat(keyValue *reflect.Value, value string) error {
	f, err := strconv.ParseFloat(value, 64)
	if err != nil {
		return createCovertionError(value, keyValue.Kind().String())
	}

	if keyValue.OverflowFloat(f) {
		return createOverflowError(value, keyValue.Kind().String())
	}

	keyValue.SetFloat(f)
	return nil
}

func setDurations(keyValue *reflect.Value, values []string) error {
	var ds = make([]time.Duration, len(values))
	for i, value := range values {
		dValue := reflect.Indirect(reflect.ValueOf(&ds[i]))
		if err := setDuration(&dValue, value); err != nil {
			return err
		}
	}
	dsValue := reflect.ValueOf(ds)
	keyValue.Set(dsValue)
	return nil
}

func setDuration(keyValue *reflect.Value, value string) error {
	duration, err := time.ParseDuration(value)
	if err != nil {
		return createCovertionError(value, "time.Duration")
	}

	durationValue := reflect.ValueOf(duration)
	keyValue.Set(durationValue)
	return nil
}

func setTimes(keyValue *reflect.Value, values []string) error {
	var ts = make([]time.Time, len(values))
	for i, value := range values {
		tValue := reflect.Indirect(reflect.ValueOf(&ts[i]))
		if err := setTime(&tValue, value); err != nil {
			return err
		}
	}
	tsValue := reflect.ValueOf(ts)
	keyValue.Set(tsValue)
	return nil
}

func setTime(keyValue *reflect.Value, value string) error {
	for _, format := range timeFormats {
		t, err := time.Parse(format, value)
		if err == nil {
			timeValue := reflect.ValueOf(t)
			keyValue.Set(timeValue)
			return nil
		}
	}

	return createCovertionError(value, "time.Time")
}

func getSectionValue(keyValue reflect.Value, sectionName string) reflect.Value {
	if sectionName == Global {
		return keyValue
	}

	sectionName = renameToPublicName(sectionName)
	return keyValue.FieldByName(sectionName)
}

// Rename the key to a public name of a struct, e.g.
//
//	"my key" -> "MyKey"
func renameToPublicName(name string) string {
	name = strings.Title(name)
	name = strings.Replace(name, " ", "", -1)
	return name
}

func getValues(value string) []string {
	values := strings.Split(value, ",")
	for i, value := range values {
		values[i] = strings.TrimSpace(value)
	}
	return values
}
